File: python/c001i.py
# Solution to /r/dailyprogrammer Challenge #1 Intermediate
# Jimmy Nguyen
#
# Prompt:
#     Create a menu driven program that will allow the user to add, edit, and
# delete events
#
# Link to challenge:
#     https://www.reddit.com/r/dailyprogrammer/comments/pihtx/intermediate_challenge_1/
def c001i():
	events = []
	command = ''
	while command != 'q':
		printEvents(events)
		printOptions(events)
		command = raw_input('\n\t>> ')
		if command == 'a':
			events = addEvent(events)
		elif command == 'e':
			events = editEvent(events)
		elif command == 'd':
			events = deleteEvent(events)

def addEvent(events):
	printEvents(events)
	print('\n\t>> add event <<\n')
	event = Event()
	event.name = raw_input('\tevent name: ')
	event.time = raw_input('\tevent time: ')
	events.append(event)
	events.sort(key=getTime)
	return events

def editEvent(events):
	printEvents(events)
	print('\n\t>> edit event <<\n')
	i = int(raw_input('\tevent number: '))-1
	print('\n\told event name: {}'.format(events[i].name))
	print('\told event time: {}\n'.format(events[i].time))
	events[i].name = raw_input('\tnew event name: ')
	events[i].time = raw_input('\tnew event time: ')
	events.sort(key=getTime)
	return events

def getTime(event):
	return event.time

def deleteEvent(events):
	printEvents(events)
	print('\n\t>> delete event <<\n')
	i = int(raw_input('\tevent number: '))-1
	del events[i]
	return events

def printOptions(events):
	print('\n\ta: add event')
	print('\td: delete event')
	if events:
		print('\te: edit event')
	print('\tq: exit')

def printEvents(events):
	print('\n==================================================\n')
	if not events:
		print('\tNo events found')
	else:
		for i in range(0,len(events)):
			printEvent(toString(i+1, len(events)), events[i])
	print('\n\t-------------------------')

def toString(val, max_val):
	val_str = ''
	for _ in range(len(str(max_val)) - len(str(val))):
		val_str += ' '
	val_str += str(val)
	return val_str

def printEvent(event_number, event):
	print('\t{}: {} - {}'.format(event_number, event.time, event.name))

class Event:
	name = None
	time = None

if __name__ == '__main__':
	c001i()

File: java/c001i.java
/*
 * Solution to /r/dailyprogrammer Challenge #1 Intermediate
 * Jimmy Nguyen
 *
 * Prompt:
 *     Create a menu driven program that will allow the user to add, edit, and
 * delete events
 *
 * Link to challenge:
 *     https://www.reddit.com/r/dailyprogrammer/comments/pihtx/intermediate_challenge_1/
 */
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public final class c001i {
	private static Scanner scanner;
	private static List<Event> events;

	private static class Event {
		public String name;
		public String time;
		public Event() {}
		public Event(String name, String time) {
			this.name = name;
			this.time = time;
		}
		public String getTime() {
			return time;
		}
	}

	private static void printEvents() {
		System.out.printf("\n==================================================\n");
		if (events.isEmpty()) {
			System.out.printf("\n\tNo events found\n");
		} else {
			for (int i = 0; i < events.size(); i++) {
				printEvent(toString(i+1, events.size()), events.get(i));
			}
		}
		System.out.printf("\n\t-------------------------\n");
	}

	private static String toString(int val, int max_val) {
		String val_str = "";
		for (int i = 0; i < String.valueOf(max_val).length() - String.valueOf(val).length(); i++) {
			val_str += " ";
		}
		val_str += String.valueOf(val);
		return val_str;
	}

	private static void printEvent(String event_number, Event event) {
		System.out.printf("\n\t%s: %s - %s", event_number, event.time, event.name);
	}

	private static void printOptions() {
		System.out.printf("\n\ta: add event");
		System.out.printf("\n\td: delete event");
		if (!events.isEmpty()) {
			System.out.printf("\n\te: edit event");
		}
		System.out.printf("\n\tq: exit");
	}

	private static String input(String prompt) {
		System.out.print(prompt);
		return scanner.nextLine();
	}

	private static void addEvents() {
		printEvents();
		System.out.printf("\n\t>> add event <<\n\n");
		Event event = new Event();
		event.name = input(String.format("\tevent name: "));
		event.time = input(String.format("\tevent time: "));
		events.add(event);
		Collections.sort(events, Comparator.comparing(Event::getTime));
	}

	private static void editEvents() {
		printEvents();
		System.out.printf("\n\t>> edit event <<\n\n");
		int i = Integer.valueOf(input(String.format("\tevent number: ")))-1;
		System.out.printf("\n\told event name: %s", events.get(i).name);
		System.out.printf("\n\told event time: %s\n\n", events.get(i).time);
		events.get(i).name = input(String.format("\tnew event name: "));
		events.get(i).time = input(String.format("\tnew event time: "));
		Collections.sort(events, Comparator.comparing(Event::getTime));
	}

	private static void deleteEvents() {
		printEvents();
		System.out.printf("\n\t>> delete event <<\n\n");
		int i = Integer.valueOf(input(String.format("\tevent number: ")))-1;
		events.remove(i);
	}

	public static void main(String[] args) {
		scanner = new Scanner(System.in);
		events = new ArrayList<Event>();
		String command = null;
		while (!"q".equals(command)) {
			printEvents();
			printOptions();
			command = input(String.format("\n\n\t>> "));
			if ("a".equals(command)) {
				addEvents();
			} else if ("e".equals(command)) {
				editEvents();
			} else if ("d".equals(command)) {
				deleteEvents();
			}
		}
	}
}
